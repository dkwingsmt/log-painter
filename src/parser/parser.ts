import foreach from 'lodash/foreach';
import values from 'lodash/values';

interface Player {
  name: string;
  colorIndex: number;
}

interface LogLine {
  time: string;
  player: Player;
  content: string;
}

interface ParseOptions {
  commandFilter: boolean;
  otherFilter: boolean;
  picTextFilter: boolean;
  timeSwitch: boolean;
  nnameSwitch: boolean;
}

interface ParseResult {
  players: Player[];
  logs: LogLine[];
}

export function parseChat(data: string, options: ParseOptions): ParseResult {
  const regHeader =
    /\d{4}-\d{2}-\d{2} (\d{1,2}:\d{2}:\d{2}) (AM|PM)? ?([^\(]*)/; // 2016-01-03 \d{2}:\d{2}:\d{2}/g name
  const regHeader2 = /(.*?)(\([0-9]+\))? +(\d{1,2}:\d{2}:\d{2}).*/; // \d{2}:\d{2}:\d{2}/
  const regDiscard =
    /^([\.。!！][rR]|[/\.、。!！][mM][eE]|[\.。!！][hH][eE][lL][pP]|[\.。!！][lL][oO][gG]|[\.。!！][wW]|[\.。!！][sS][cC]|[\.。!！][eE][nN]|[\.。!！][sS][eE][tT]|[\.。!！][sS][tT]|[\.。!！][cC][oO][cC]|[\.。!！][dD][nN][dD]|[\.。!！][tTlL][iI]|[\.。!！][jJ][rR][rR][pP]|[\.。!！][rR][uU][lL][eE][sS]|[\.。!！][nN]|[\.。!！][bB][oO][tT]|[\.。!！][oO][bB]|[\.。!！][wW][eE][lL][cC][oO][mM][eE]).*/;
  const regOther = /^(\(|（).*/;
  const regPicText = /\[图片\]/;

  const nameList: Record<string, Player> = {};
  const logList: LogLine[] = [];

  let colorIndex = 0;

  function ensureAddlayer(name: string): void {
    if (!(name in nameList)) {
      nameList[name] = { name, colorIndex: colorIndex };
      colorIndex++;
    }
  }

  let name = "";
  let time = "";

  foreach(data.split("\n"), (line) => {
    if (line.length == 0) {
      return;
    }
    // Recognize message header
    const res = line.match(regHeader);
    const res2 = line.match(regHeader2);

    // Deal with new name

    if (res != null || res2 != null) {
      if (res != null) {
        name = res[3].trim();
        time = res[1];
      } else if (res2 != null) {
        name = res2[1].trim();
        time = res2[3];
      }
      //----------------NEW NAME
      if (options.nnameSwitch && name.indexOf('【') == 0 && name.indexOf('】') != -1) {
        name = name.substr(name.indexOf('】') + 1);
      }
      //----------------------------------------------
      if (name != "" && name != "系统消息") {
        ensureAddlayer(name);
      }

    } else {
      const player: Player | undefined = nameList[name];
      if (player == null) {
        return;
      }

      //Deal with message content
      if (options.commandFilter) {
        const discard = line.match(regDiscard);
        if (discard != null) {
          return;
        }
      }

      if (options.otherFilter) {
        const other = line.match(regOther);
        if (other != null) {
          return;
        }
      }

      let editedLine = line;
      if (options.picTextFilter) {
        editedLine = line.replace(regPicText, "");
        if (line.length == 0) {
          return;
        }
      }

      //Fix time length to 8
      if (time.length != 8) {
        time = ' ' + time;
      }

      const log: LogLine = { time, player, content: editedLine };
      logList[logList.length] = log;
    }
  });

  return { players: values(nameList), logs: logList };
}