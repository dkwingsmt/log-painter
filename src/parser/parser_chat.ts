import compact from 'lodash/compact';
import trim from 'lodash/trim';
import flow from 'lodash/flow';

export interface ParsedPlayer {
  name: string;
  number?: string;
  title?: string;
}

export interface ParsedLine {
  time?: string;
  player: ParsedPlayer;
  content: string[];
}

export interface ParseResult {
  logLines: ParsedLine[];
}

interface ParsedHeader {
  player: ParsedPlayer;
  time?: string;
}

type HeaderParser = (line: string) => ParsedHeader | null;
type LogLineConverter = (logLine: ParsedLine) => ParsedLine | null;

interface LogConfig {
  headerParser: HeaderParser;
  logLineConverter: LogLineConverter;
}

function removeSystemTextConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const withdrawParser = /^.*撤回了一条消息( 重新编辑)?$/;
  const withdrawMemberParser = /^.*撤回了成员.*的一条消息$/;
  const withdrawMember2Parser = /^.*撤回了一条成员消息$/;
  const commonFriendsParser = /^你和.*有\d+个共同好友，点击添加好友。$/;
  const joinMemberParser = /^.*加入本群。$/;
  const inviteMemberParser = /^.*邀请.*加入了本群。$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      withdrawParser,
      withdrawMemberParser,
      withdrawMember2Parser,
      commonFriendsParser,
      joinMemberParser,
      inviteMemberParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

const regNumber = /\(\d+\)|<.+@.+\..+>/;
const regTitle = /(?:【(.{1,6})】)?/;
const regTime = /\d{1,2}:\d{2}:\d{2}(?: (?:AM|PM))?/;

function removeMessageManagerSystemTextConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const dateParser = /^ \d{4}-\d{2}-\d{2}$/;
  const withdrawParser = new RegExp(`^${regTime.source}.*撤回了一条消息$`);
  const withdrawMemberParser = new RegExp(`^${regTime.source}.*撤回了成员.*的一条消息$`);
  const withdrawMember2Parser = new RegExp(`^${regTime.source}.*撤回了一条成员消息$`);
  // const commonFriendsParser = /^你和.*有\d+个共同好友，点击添加好友。$/;
  const joinMemberParser = new RegExp(`^${regTime.source}.*加入本群。$`);
  // const inviteMemberParser = /^.*邀请.*加入了本群。$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      dateParser,
      withdrawParser,
      withdrawMemberParser,
      withdrawMember2Parser,
      joinMemberParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

function defaultConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;
  // Remove starting and trailing empty lines.
  while(lines.length > 0 && lines[0].trim() === '') {
    lines.splice(0, 1);
  }
  while(lines.length > 0 && lines[lines.length-1].trim() === '') {
    lines.splice(lines.length-1, 1);
  }
  if (lines.length === 0)
    return null;
  return logLine;
}

function trimNumber(src: string): string {
  return trim(src, '()<>');
}

// Export from log
// E.g. "2019-09-23 8:43:38 PM 骰娘-Roll100(872001750)"
const exportFromLog: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regDateName = new RegExp(`\\d{4}-\\d{2}-\\d{2} ${regTime.source}`);
    const regHeader = new RegExp(`^(${regDateName.source}) (.*)(${regNumber.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, time, name, number] = matches;
    return {
      player: {
        name,
        number: trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    if (['10000', '1000000'].includes(logLine.player.number || ''))
      return null;
    return defaultConverter(logLine);
  },
};

// Copy from sidewindow
// E.g. "【冒泡】无情的围观熊 2/14/2020 9:16:13 PM"
const copyFromSideWindow: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regDateTime = new RegExp(`\\d{1,4}\\/\\d{1,2}\\/\\d{1,4} ${regTime.source}`);
    const regHeader = new RegExp(`^${regTitle.source}(.*?) (${regDateTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, time] = matches;
    return {
      player: {
        name,
        title,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from chat
// E.g. "【煤油】丧 丧 熊 9:59:54 PM"
const copyFromChat: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = new RegExp(`^${regTitle.source}(.*) (${regTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, time] = matches;
    return {
      player: {
        name,
        title,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from message manager
// E.g. "织练取(958884) 3:13:28 AM"
const copyFromMessageManager: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = new RegExp(`^${regTitle.source}(.*?)(${regNumber.source}) (${regTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, number, time] = matches;
    return {
      player: {
        name,
        title,
        number: trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeMessageManagerSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from mobile
// E.g. "a dark ideation  23:50:40"
const copyFromMobile: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = /^(.*?) {2}(\d{1,2}:\d{2}:\d{2}) *$/;
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, time] = matches;
    return {
      player: {
        name,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return defaultConverter(logLine);
  },
};

export function parseChat(data: string): ParseResult {
  const logLines: ParsedLine[] = [];
  let firstLogConfig: LogConfig | undefined = undefined;
  for (const line of data.split('\n')) {
    const parsedHeader = ((): ParsedHeader | null => {
      if (firstLogConfig)
        return firstLogConfig.headerParser(line);
      for (const logConfig of [
        copyFromMessageManager,
        copyFromMobile,
        exportFromLog,
        copyFromSideWindow,
        copyFromChat,
      ]) {
        const result = logConfig.headerParser(line);
        if (result) {
          firstLogConfig = logConfig;
          return result;
        }
      }
      return null;
    })();
    if (parsedHeader) {
      logLines.push({
        ...parsedHeader,
        content: [],
      });
    } else {
      if (firstLogConfig) {
        const currentLogLine: ParsedLine | undefined = logLines[logLines.length - 1];
        // Content after a recognizable header
        if (currentLogLine)
          currentLogLine.content.push(line);
      } else {
        // Unrecognizable content before the first header
        // Do nothing
      }
    }
  }
  if (!firstLogConfig)
    return { logLines: [] };

  const convertedLogLines = compact(
    logLines.map(firstLogConfig.logLineConverter)
  );
  return { logLines: convertedLogLines };
}