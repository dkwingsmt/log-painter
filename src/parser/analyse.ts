import forEach from 'lodash/forEach';
import values from 'lodash/values';
import clone from 'lodash/clone';
import { ParseResult, ParsedPlayer } from './parser_chat';
import {
  Configuration,
  AnalysedLine,
  AnalysedPlayer,
  AnalyseResult,
  ConfigPlayer,
  presetDescribedColors,
  DescribedColor,
} from 'common';

interface GroupResult {
  players: AnalysedPlayer[];
  lines: AnalysedLine[];
};

const presetColors: string[] = presetDescribedColors.map(({ value }: DescribedColor) => value);

function getPlayerIdGroup(player: ParsedPlayer): string[] {
  const ids: string[] = [];
  if (player.number)
    ids.push(`qq:${player.number}`);
  ids.push(`name:${player.name}`);
  return ids;
}

function groupByPlayers(result: ParseResult): GroupResult {
  const lines: AnalysedLine[] = [];
  const players: Record<string, AnalysedPlayer> = {};
  for (const parsedLine of result.logLines) {
    const { player, time, content } = parsedLine;
    const playerIdGroup = getPlayerIdGroup(player);
    const majorPlayerId = playerIdGroup[0];
    if (!players[majorPlayerId])
      players[majorPlayerId] = {
        playerId: majorPlayerId,
        allPlayerIds: playerIdGroup,
        name: player.name,
        number: player.number,
      };
    lines.push({
      playerId: majorPlayerId,
      time,
      title: player.title,
      content,
    });
  }
  return {
    lines,
    players: values(players),
  };
}

function findFirstPlayerConfig(
  configPlayers: Record<string, ConfigPlayer>,
  playerIdGroup: string[],
): ConfigPlayer | null {
  for (const playerId of playerIdGroup) {
    if (configPlayers[playerId])
      return configPlayers[playerId];
  }
  return null;
}

function findFirstUnusedPresetColor(
  presetColors: string[],
  configPlayers: Record<string, ConfigPlayer>,
): string {
  const existingColors: Record<string, boolean> = {};
  forEach(configPlayers, (configPlayer: ConfigPlayer) => {
    existingColors[configPlayer.color] = true;
  });
  return presetColors.find((color: string) => !existingColors[color]) || 'black';
}

export function analyse(parsed: ParseResult, currentConfig: Configuration): AnalyseResult {
  const { lines, players } = groupByPlayers(parsed);
  const configPlayers: Record<string, ConfigPlayer> = clone(currentConfig.players);
  for (const player of players) {
    let configPlayer = findFirstPlayerConfig(configPlayers, player.allPlayerIds);
    if (!configPlayer) {
      configPlayer = {
        enabled: true,
        displayName: player.name,
        color: findFirstUnusedPresetColor(presetColors, configPlayers),
      };
    }
    for (const playerId of player.allPlayerIds) {
      configPlayers[playerId] = configPlayer;
    }
  };
  return {
    lines,
    playerIds: players.map((player: AnalysedPlayer) => player.playerId),
    nextConfig: {
      players: configPlayers,
    },
  };
}