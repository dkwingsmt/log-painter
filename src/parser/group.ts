import values from 'lodash/values';

import { ParseResult, ParsedPlayer } from './parser_chat';
import {
  GroupResult,
  AnalysedLine,
  AnalysedPlayer,
} from 'common';

function getPlayerIdGroup(player: ParsedPlayer): string[] {
  const ids: string[] = [];
  if (player.number)
    ids.push(`qq:${player.number}`);
  ids.push(`name:${player.name}`);
  return ids;
}

export function groupByPlayers(result: ParseResult): GroupResult {
  const lines: AnalysedLine[] = [];
  const players: Record<string, AnalysedPlayer> = {};
  for (const parsedLine of result.logLines) {
    const { player, time, content } = parsedLine;
    const playerIdGroup = getPlayerIdGroup(player);
    const majorPlayerId = playerIdGroup[0];
    if (!players[majorPlayerId])
      players[majorPlayerId] = {
        playerId: majorPlayerId,
        allPlayerIds: playerIdGroup,
        name: player.name,
        number: player.number,
      };
    lines.push({
      playerId: majorPlayerId,
      time,
      title: player.title,
      content,
    });
  }
  return {
    lines,
    players: values(players),
  };
}
