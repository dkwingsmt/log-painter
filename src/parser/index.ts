import { analyse, AnalyseResult } from './analyse';
import { parseChat, ParseResult } from './parser_chat';
import { Configuration } from 'common';

export function convert(source: string, currentConfig: Configuration): AnalyseResult {
  const parseResult: ParseResult = parseChat(source);
  return analyse(parseResult, currentConfig);
}