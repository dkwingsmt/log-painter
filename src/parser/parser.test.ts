import { parseChat } from "parser";

it("basic", () => {
  const text = `
消息记录（此消息记录为文本格式，不支持重新导入）

================================================================
消息分组:我的群聊
================================================================
消息对象:湖之仆从（PST 8pm | BJT 11am）
================================================================

2019-09-23 8:43:38 PM 骰娘-Roll100(872001750)
1

2019-09-23 8:43:38 PM 系统消息(10000)
[图片]2

2019-09-23 8:44:09 PM 系统消息(1000000)
3

2019-09-23 8:46:18 PM 暗夜(280848250)


2019-09-23 8:49:12 PM 系统消息(1000000)
真核生物独幕剧加入本群。

2019-09-23 9:02:42 PM 暗夜(280848250)
.r test

2019-09-23 11:26:00 PM 岁寒(362264070)
4

2019-09-24 7:54:29 AM 黄山真君(1065057234)
5

2019-09-24 9:16:35 AM 黄山真君(1065057234)
6
  `;
  expect(
    parseChat(text, {
      commandFilter: false,
      otherFilter: false,
      picTextFilter: false,
      timeSwitch: false,
      nnameSwitch: false
    })
  ).toMatchSnapshot();
});
