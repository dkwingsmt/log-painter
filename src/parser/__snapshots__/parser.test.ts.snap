// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
Object {
  "logs": Array [
    Object {
      "content": "1",
      "player": Object {
        "colorIndex": 0,
        "name": "骰娘-Roll100",
      },
      "time": " 8:43:38",
    },
    Object {
      "content": ".r test",
      "player": Object {
        "colorIndex": 1,
        "name": "暗夜",
      },
      "time": " 9:02:42",
    },
    Object {
      "content": "4",
      "player": Object {
        "colorIndex": 2,
        "name": "岁寒",
      },
      "time": "11:26:00",
    },
    Object {
      "content": "5",
      "player": Object {
        "colorIndex": 3,
        "name": "黄山真君",
      },
      "time": " 7:54:29",
    },
    Object {
      "content": "6",
      "player": Object {
        "colorIndex": 3,
        "name": "黄山真君",
      },
      "time": " 9:16:35",
    },
    Object {
      "content": "  ",
      "player": Object {
        "colorIndex": 3,
        "name": "黄山真君",
      },
      "time": " 9:16:35",
    },
  ],
  "players": Array [
    Object {
      "colorIndex": 0,
      "name": "骰娘-Roll100",
    },
    Object {
      "colorIndex": 1,
      "name": "暗夜",
    },
    Object {
      "colorIndex": 2,
      "name": "岁寒",
    },
    Object {
      "colorIndex": 3,
      "name": "黄山真君",
    },
  ],
}
`;
